source/Test.cpp

```diff
@@ -0,0 +1,100 @@
+/* Test.cpp
+Copyright (c) 2019 by Peter van der Meer
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+
+#include "DataNode.h"
+#include "PlayerInfo.h"
+#include "Test.h"
+#include "TestStep.h"
+#include "UI.h"
+#include <string>
+
+using namespace std;
+
+
+
+Test::Test()
+{
+	name = "";
+}
+
+
+
+Test::~Test()
+{
+}
+
+
+
+void Test::Load(const DataNode &node)
+{
+	if(node.Size() < 2)
+	{
+		node.PrintTrace("No name specified for test");
+		return;
+	}
+	if (node.Token(0) != "test")
+	{
+		node.PrintTrace("Non-test found in test parsing");
+		return;
+	}
+	name = node.Token(1);
+
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "status" && child.Size() >= 2)
+		{
+			if (child.Token(1) == "Active")
+				status = STATUS_ACTIVE;
+			else if (child.Token(1) == "Known Failure")
+				status = STATUS_KNOWN_FAILURE;
+			else if (child.Token(1) == "Missing Feature")
+				status = STATUS_MISSING_FEATURE;
+		}
+		else if (child.Token(0) == "sequence")
+			for (const DataNode &seqChild : child)
+			{
+				testSteps.push_back(new TestStep());
```

Comment:
Do we need to use `new`? For every `new` there must be a corresponding `delete` lest we leak memory. The std::vector destructor will not call delete on a raw pointer for you.

I would imagine there is a way we can use emplace_back with the node and a TestStep constructor that takes a DataNode to simplify this and the next line.

Reply:
Yes, emplace_back is probably better here. I'm also considering to move TestStep to become a nested class in Test. It looks like the TestStep class doesn't really need to be a stand-alone class.

Reply:
Changed to emplace_back